### ISSUE TITLE
AuditSubscriber injects a LoggerChain into a LoggerChain when loading fixtures

### ISSUE THREAD

**Issue by acanicatti** [DamienHarper/DoctrineAuditBundle#70]
When loading doctrine fixtures, the following code in the onFlush method of the AuditSubscriber seems to be buggy:

```
$this->loggerBackup = $em->getConnection()->getConfiguration()->getSQLLogger();
$loggerChain = new LoggerChain();
$loggerChain->addLogger(new AuditLogger(function () use ($em) {
    // flushes pending data
    $em->getConnection()->getConfiguration()->setSQLLogger($this->loggerBackup);
    $uow = $em->getUnitOfWork();

     $this->manager->processInsertions($em, $uow);
     $this->manager->processUpdates($em, $uow);
     $this->manager->processAssociations($em);
     $this->manager->processDissociations($em);
     $this->manager->processDeletions($em);

     $this->manager->reset();
}));

if ($this->loggerBackup instanceof SQLLogger) {
    $loggerChain->addLogger($this->loggerBackup);
}

$em->getConnection()->getConfiguration()->setSQLLogger($loggerChain);
```

The `loggerBackup` variable is of type `Doctrine\DBAL\Logging\LoggerChain` (which implements SQLLogger). In the code above a new LoggerChain is created and an AuditLogger is added to it. But then the loggerBackup is also added to the new LoggerChain since LoggerChain implements SQLLogger. So now we have a LoggerChain inside a LoggerChain.
Below is the `startQuery` method of the LoggerChain:
```    
/**
 * {@inheritdoc}
 */
public function startQuery($sql, array $params = null, array $types = null)
{
    foreach ($this->loggers as $logger) {
        $logger->startQuery($sql, $params, $types);
    }
}
```
The method will recurse infinitely, and with xdebug enabled, you'll get a max_nesting_level error. Not exactly sure if this issue is being caused by something else, or if I'm missing something, but these are my initial findings. 

Any and all help is greatly appreciated.

**Comment by DamienHarper** [comment]
Hi @acanicatti. Thanks for your feedback.

I think something else is the root cause of your infinite loop, let me try to explain your snippet

>The loggerBackup variable is of type Doctrine\DBAL\Logging\LoggerChain (which implements SQLLogger). 

Right, it it used to backup the current logger instance of the doctrine connection. It may be a `LoggerChain` instance, let's call it `LoggerChain#123`

> In the code above a new LoggerChain is created and an AuditLogger is added to it. 

Right, we create a blank `LoggerChain` instance (let's call it `LoggerChain#456`) then add to it an `AuditLogger` instance (let's call it `AuditLogger#135`).

> But then the loggerBackup is also added to the new LoggerChain since LoggerChain implements SQLLogger. So now we have a LoggerChain inside a LoggerChain.

Well we add to the newly created `LoggerChain` instance the one we backed up at the beginning.
So first we add `AuditLogger#135` to `LoggerChain#456`, then we add the backed up `LoggerChain#123` to `LoggerChain#456`

Finally:
* We have a `LoggerChain` instance inside another `LoggerChain` instance (`LoggerChain#123` inside `LoggerChain#456`)
* The backed up `LoggerChain` instance (`LoggerChain#123`) is a different instance than the newly one (`LoggerChain#456`).
* `LoggerChain#123` do not have any (cyclic) reference to `LoggerChain#456`

When `LoggerChain::startQuery()` is run, it iterates overs embedded logger instances and call a `startQuery` on each.
1. it loops over `LoggerChain#456` logger instances (there are two)
1.1. it calls `AuditLogger#135::startQuery()`
1.2. it calls `LoggerChain#123::startQuery()` which in turn loops over its logger instances etc

**Comment by acanicatti** [comment]
Hi Damien,

Yep, I totally understand what you're saying. My issue was that the call stack was too big because my project loads many many fixtures - which are flushed in one transaction at the end of the command. So yes, the `LoggerChain::startQuery()` method is looping over logger instances that are all different objects, but it was doing it too much and getting the max_nesting_level error. 

I've read elsewhere that this error only happens with xdebug enabled, which I only use in dev environments, so I don't see it being an issue. I think its perfectly valid to increase the max_nesting_level parameter in my xdebug config, provided my application is loading so many fixtures.

Thanks so much for your help!
Anthony

**Comment by bzoks** [comment]
Hi!
I have similar issue, but didn't had time yet to debug it and figure our reasons...
In my case I execute hundreds of inserts and then issue a commit... and hit a `maximum function nesting level...` from xdebug, even `with xdebug.max_nesting_level = 1000`

**Comment by acanicatti** [comment]
@bzoks agreed. I think the simplest solution would be for the AuditSubscriber to detect if the current SQLLogger is a LoggerChain and "flatten" it by appending its children to the newly created LoggerChain. This way a LoggerChain could never exist inside another LoggerChain. (I'm not sure if this has side effects, but I don't see any reason the chain of SQLLoggers can be one flat chain as opposed to nested LoggerChains).

Being that its only an issue on dev environments I don't mind modifying the code by hand but I would also appreciate a real fix.

Anthony

**Comment by DamienHarper** [comment]
@acanicatti @bzoks I reopened the issue to keep it in sight. 
I was thinking of something like what @acanicatti suggested: flattening loggers if possible. I'll dig into it asap.
Anyway, feel free to submit a PR, I'd be happy to review/merge it ;)

**Comment by acanicatti** [comment]
Will do as soon as I get the chance! Thanks again for the help!

Anthony

**Comment by acanicatti** [comment]
Not sure if this is a related issue but now I'm having a problem with this code in the `AuditHelper` class:
```
public function summarize(EntityManager $em, $entity = null, $id = null): ?array
{
    if (null === $entity) {
        return null;
    }
     $em->getUnitOfWork()->initializeObject($entity); // ensure that proxies are initialized
    $meta = $em->getClassMetadata(\get_class($entity));
    $pkName = $meta->getSingleIdentifierFieldName();
    $pkValue = $id ?? $this->id($em, $entity);

   if (method_exists($entity, '__toString')) {
        $label = (string) $entity;
   } else {
        $label = \get_class($entity).'#'.$pkValue;
   }

   return [
        'label' => $label,
        'class' => $meta->name,
        'table' => $meta->getTableName(),
        $pkName => $pkValue,
   ];
}
```
For whatever reason one of my entities is not being initialized properly and the `id` function is returning null. So the `pkValue` var is null and casting to a string throws an Exception. The Proxy object is clearly not initialized but calling the `initialize` function doesn't seem to initialize it fully either. 

Would you have any idea why that might be happening? I can add a null check in the method to prevent the error but I'd prefer to fully understand why it happens as well. 

Thanks
Anthony

**Comment by DamienHarper** [comment]
@acanicatti Could you provide me some concrete code (either as attachments or as a private gist) that reproduce the issue? it would be very helpful

**Comment by DamienHarper** [comment]
Fixed by #72
